src/protocol/ipa/mod.rs

```diff
@@ -771,132 +771,195 @@ pub mod tests {
-                        &input_rows,
-                        per_user_cap,
-                        MAX_BREAKDOWN_KEY as u128,
-                        NUM_MULTI_BITS,
-                    )
-                    .await
-                    .unwrap()
-                })
+        let result: Vec<GenericReportTestInput<Fp32BitPrime, MatchKey, BreakdownKey>> = world
+            .semi_honest(records, |ctx, input_rows| async move {
+                ipa::<Fp32BitPrime, MatchKey, BreakdownKey>(
+                    ctx,
+                    &input_rows,
+                    per_user_cap,
+                    max_breakdown_key as u128,
+                    NUM_MULTI_BITS,
+                )
                 .await
-                .reconstruct();
+                .unwrap()
+            })
+            .await
+            .reconstruct();
 
-            assert_eq!(MAX_BREAKDOWN_KEY, result.len());
-            println!(
-                "actual results: {:#?}",
-                result
-                    .iter()
-                    .map(|x| x.trigger_value.as_u128())
-                    .collect::<Vec<_>>(),
+        assert_eq!(max_breakdown_key, result.len());
+        println!(
+            "actual results: {:#?}",
+            result
+                .iter()
+                .map(|x| x.trigger_value.as_u128())
+                .collect::<Vec<_>>(),
+        );
+        for (i, expected) in expected_results.iter().enumerate() {
+            assert_eq!(
+                [i as u128, u128::from(*expected)],
+                [
+                    result[i].breakdown_key.as_u128(),
+                    result[i].trigger_value.as_u128()
+                ]
             );
-            for (i, expected) in expected_results.iter().enumerate() {
-                assert_eq!(
-                    [i as u128, u128::from(*expected)],
-                    [
-                        result[i].breakdown_key.as_u128(),
-                        result[i].trigger_value.as_u128()
-                    ]
+        }
+    }
+
+    #[tokio::test]
+    #[allow(clippy::missing_panics_doc)]
+    pub async fn random_ipa_check() {
+        const MAX_BREAKDOWN_KEY: usize = 16;
+        const MAX_TRIGGER_VALUE: u32 = 5;
+        const NUM_USERS: usize = 10;
+        const MAX_RECORDS_PER_USER: usize = 8;
+        const NUM_MULTI_BITS: u32 = 3;
+
+        let random_seed = thread_rng().gen();
+        println!("Using random seed: {random_seed}");
+        let mut rng = StdRng::seed_from_u64(random_seed);
+
+        let mut random_user_records = Vec::with_capacity(NUM_USERS);
+        for _ in 0..NUM_USERS {
+            let records_for_user = generate_random_user_records_in_reverse_chronological_order(
+                &mut rng,
+                MAX_RECORDS_PER_USER,
+                MAX_BREAKDOWN_KEY,
+                MAX_TRIGGER_VALUE,
+            );
+            random_user_records.push(records_for_user);
+        }
+        let mut raw_data = random_user_records.concat();
```

Comment:
I would use `Vec::append()` for each additional user rather than building two vectors.

Reply:
The good reason to do this, is that I can pull the random test data *outside* of the for loop instead of re-generating it each time. But since the expected result changes when you modify per_user_cap, I need to recalculate that, and if I've already appended all of the records from all the users together, that's a lot harder to do.